File: arc/sources/x.py
"""X / Twitter — Bearer token required. Stubs gracefully if absent."""
from __future__ import annotations

import requests

from arc.sources.base import Source, RawTrend
from arc.config import X_BEARER_TOKEN

_API_BASE = "https://api.twitter.com/2"


class XSource(Source):

    def fetch(self) -> list[RawTrend]:
        if not X_BEARER_TOKEN:
            print("[x] X_BEARER_TOKEN not set — skipping X source")
            return []

        query = self.handle or "AI tools"
        return self._search_recent(query)

    def _search_recent(self, query: str) -> list[RawTrend]:
        """Twitter v2 recent search — free tier allows 500k tweets/month."""
        params = {
            "query": f"{query} -is:retweet lang:en",
            "max_results": 20,
            "tweet.fields": "created_at,public_metrics,author_id",
        }
        headers = {"Authorization": f"Bearer {X_BEARER_TOKEN}"}

        try:
            res = requests.get(
                f"{_API_BASE}/tweets/search/recent",
                params=params,
                headers=headers,
                timeout=10,
            )
            if res.status_code == 403:
                print("[x] 403 — free tier doesn't include search. Skipping.")
                return []
            res.raise_for_status()
            data = res.json().get("data", [])
        except Exception as e:
            print(f"[x] fetch failed: {e}")
            return []

        results: list[RawTrend] = []
        for tweet in data:
            tid = tweet.get("id", "")
            text = tweet.get("text", "").strip()
            if not text:
                continue

            metrics = tweet.get("public_metrics", {})
            likes = metrics.get("like_count", 0)
            retweets = metrics.get("retweet_count", 0)
            velocity = float(likes + retweets * 2)   # simple engagement score

            results.append(RawTrend(
                external_id=f"x-{tid}",
                title=text[:200],
                url=f"https://x.com/i/web/status/{tid}",
                velocity=velocity,
                raw={
                    "source": "x",
                    "metrics": metrics,
                    "created_at": tweet.get("created_at", ""),
                },
            ))

        return results

File: src/lib/arc/sources/x.ts
import type { RawTrend, SourceRow } from "./index";

export async function fetchX(row: SourceRow): Promise<RawTrend[]> {
  const token = process.env.X_BEARER_TOKEN;
  if (!token) {
    console.log("[x] X_BEARER_TOKEN not set, skipping");
    return [];
  }

  const query = row.handle ?? "AI tools";
  const url =
    `https://api.twitter.com/2/tweets/search/recent` +
    `?query=${encodeURIComponent(query + " -is:retweet lang:en")}` +
    `&max_results=20&tweet.fields=created_at,public_metrics`;

  try {
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
      signal: AbortSignal.timeout(10000),
    });

    if (res.status === 403) {
      console.log("[x] 403, free tier doesn't include search");
      return [];
    }
    if (!res.ok) {
      console.error(`[x] HTTP ${res.status}`);
      return [];
    }

    const json = await res.json();
    return (json.data ?? []).map((tweet: {
      id: string;
      text: string;
      created_at?: string;
      public_metrics?: { like_count: number; retweet_count: number };
    }) => {
      const m = tweet.public_metrics ?? { like_count: 0, retweet_count: 0 };
      const velocity = m.like_count + m.retweet_count * 2;

      return {
        external_id: `x-${tweet.id}`,
        title: tweet.text.slice(0, 200),
        url: `https://x.com/i/web/status/${tweet.id}`,
        velocity,
        raw: { source: "x", metrics: m, created_at: tweet.created_at ?? "" },
      };
    });
  } catch (e) {
    console.error("[x] fetch failed:", e);
    return [];
  }
}
